compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -1017,7 +1018,31 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             return Ok(cycle_result);
         }
 
-        let (result, dep_node) = self.in_task(|this| this.evaluate_stack(&stack));
+        let (result, dep_node) = self.in_task(|this| {
+            let mut result = this.evaluate_stack(&stack)?;
+
+            // if evaluate failed, try to normalize the predicate and evaluate again.
+            if EvaluationResult::EvaluatedToErr == result
```

Comment:
this doesn't feel like the right place to do this. The caller of evaluate should ideally have already normalized the predicate.

I feel like we're missing a normalize call somewhere and that this change is merely hiding that issue.

Haven't read the discussion in this PR so this may have already been brought up, sorry in that case.

Reply:
alright :sweat_smile: actually looked at the previous comments and saw this already mentioned. We intend to replace evaluate with a new solver in the near future at which point this issue should be fixed without needing any additional work.

Until then I am always a bit concerned when making evaluate stronger as this provides more places where the specific behavior of evaluate is actually more powerful than the new solver making it more difficult to switch. I don't think this PR adds any such issues cause it should just work™